examples/autoignition-mpi.py

```diff
@@ -240,71 +308,135 @@ def main(ctx_factory=cl.create_some_context, casename="autoignition", use_leap=F
-                              viz_fields=viz_fields)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                checkpoint=my_checkpoint,
-                get_timestep=get_timestep, state=current_state,
-                t=current_t, t_final=t_final)
-    except ExactSolutionMismatch as ex:
-        error_state = True
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
-
-    if not check_step(current_step, nviz):  # If final step not an output step
-        if rank == 0:
-            logger.info("Checkpointing final state ...")
-        my_checkpoint(current_step, t=current_t,
-                      dt=(current_t - checkpoint_t),
-                      state=current_state)
-
-    if current_t - t_final < 0:
-        error_state = True
-
-    if error_state:
-        raise ValueError("Simulation did not complete successfully.")
+    def my_post_step(step, t, dt, state):
+        # Logmgr needs to know about EOS, dt, dim?
+        # imo this is a design/scope flaw
+        if logmgr:
+            set_dt(logmgr, dt)
+            set_sim_state(logmgr, dim, state, eos)
+            logmgr.tick_after()
+        return state, dt
+
+    def my_pre_step(step, t, dt, state):
+        try:
+            dv = None
+
+            if logmgr:
+                logmgr.tick_before()
+
+            from mirgecom.simutil import check_step
+            do_viz = check_step(step=step, interval=nviz)
+            do_restart = check_step(step=step, interval=nrestart)
+            do_health = check_step(step=step, interval=nhealth)
+
+            if step == rst_step:  # don't do viz or restart @ restart
+                do_viz = False
+                do_restart = False
+
+            if do_health:
+                dv = eos.dependent_vars(state)
+                health_errors = my_health_check(dv)
+                if comm is not None:
+                    health_errors = comm.allreduce(health_errors, op=MPI.LOR)
+                if health_errors:
+                    if rank == 0:
+                        logger.info("Fluid solution failed health check.")
+                    raise HealthCheckError()
+
+            if do_restart:
+                my_write_restart(step=step, t=t, state=state)
+
+            if do_viz:
+                production_rates = eos.get_production_rates(state)
+                if dv is None:
+                    dv = eos.dependent_vars(state)
+                my_write_viz(step=step, t=t, state=state, dv=dv,
+                             production_rates=production_rates)
+
+        except MyError:
+            if rank == 0:
+                logger.info("Errors detected; attempting graceful exit.")
+            my_write_viz(step=step, t=t, state=state)
+            my_write_restart(step=step, t=t, state=state)
+            raise
+
+        t_remaining = max(0, t_final - t)
+        return state, min(dt, t_remaining)
```

Comment:
Do we want to use `get_inviscid_timestep` here too, now that `get_timestep` is gone? Or are we leaving that for #368?

Reply:
This hints at some aspects of what was wrong with `get_timestep` and `get_inviscid_timestep`.  `get_inviscid_timestep` is only useful if DT is going to change, adapting to the solution.  In all of our current examples DT is constant. So it is completely unneeded.  The current `get_inviscid_timestep` in main is broken.  #368 fixes it.

This little chunk of code that you've highlighted is all that is needed currently to ensure that the users' constant DT is respected up to where it closes in on `t_final`, and that timestepping does not surpass `t_final`.

As a preview, in #368, we need to check whether we are doing constant DT, or constant CFL, and calculate DT (or CFL) accordingly.  If the user wants, he can just use `get_sim_timestep` built-in utility that will automatically do the right thing.  On the other hand, if the user wants to report the current CFL through logging or otherwise, or visualize the point-local CFL or DT, then they need to handle that in the callback because `get_sim_timestep` does not offer an interface to return the point-local fields. 

Does this make sense so far?

Reply:
Incidentally @majosm, the finishing assert at the end of the driver nicely caught the issue and reminded me that upon eliminating the `get_timestep` junk, that this DT snippet needed to be added to the callback.  This particular snippet and approach has been reviewed in-person with AK.

Reply:
I guess I was mainly just wondering about the `nel_1d` and `order` scaling factors in there, whether we wanted to retain that functionality somehow for the time being until #368 (either by calling `get_inviscid_timestep` or just pasting the expression into the `current_dt = ...` in the driver?).

Reply:
I would say no. They are unused as-is, and worse, they are *wrong*. These have been replaced even in wave-eager down inside grudge - favoring the developments similar to those in #368.

We have constant DT, and the timestep is never recalculated, which makes even calling these functions from the driver misleading.